enhancements/ingress/set-replace-append-delete-deny-headers.md

```diff
@@ -0,0 +1,603 @@
+The Ingress Controller API is extended by adding an optional `httpSetHeaders` with type 
+`IngressControllerSetHTTPHeaders` to `IngressControllerSpec`.
+
+```go
+type IngressControllerSpec struct {
+  // 
+  // httpSetHeaders defines HTTP headers that should be set(replace)/append.
+  // If this field is empty, no headers are added.
+  //
+  // Note that this option only applies to cleartext HTTP connections
+  // and to secure HTTP connections for which the ingress controller
+  // terminates encryption (that is, edge-terminated or reencrypt
+  // connections).  Headers cannot be set(replace)/append for TLS passthrough
+  // connections.
+  //
+  // +kubebuilder:validation:Optional
+  // +optional
+  HTTPSetHeaders IngressControllerSetHTTPHeaders `json:"httpSetHeaders,omitempty"`
+}
+```
+
+The IngressControllerSetHTTPHeaders type has a Response field,
+of type []IngressControllerSetHTTPHeader, for specifying headers to set/append:
+
+```go
+// The IngressControllerSetHTTPHeaders type has a Response field,
+// type []IngressControllerSetHTTPHeader, for specifying headers to set(replace) or append:
+
+type IngressControllerSetHTTPHeaders struct {
+  // response specifies which HTTP response headers to set(replace) or append.
+  //
+  // If this field is empty, no response headers are set(replace) or append.
+  //
+  // +nullable
+  // +optional
+  Response []IngressControllerSetHTTPHeader `json:"response,omitempty"`
+}
+```
+
+The IngressControllerSetHTTPHeader type has fields for specifying a header name, header value and action.
+```go
+type IngressControllerSetHTTPHeader struct{
+    // name specifies a header name.  Its value must be a valid HTTP header 
+    // name as defined in RFC 2616 section 4.2.
+    //	
+    // +kubebuilder:validation:Required
+    // +kubebuilder:validation:MinLength=1
+    // +kubebuilder:validation:MaxLength=256
+    // +kubebuilder:validation:Pattern="^[-!#$%&'*+.0-9A-Z^_`a-z|~]+$" 
+    // +required 
+    Name string `json:"name"`
+
+    // Value specifies a header value.
+    // +kubebuilder:validation:Required 
+    // +required
+    // +kubebuilder:validation:MinLength=1
+    // +kubebuilder:validation:MaxLength=4096
+    Value string `json:"value"`
+
+    // Action specifies a header value.
+    // +kubebuilder:validation:Required
+    // +required
+    // +kubebuilder:validation:Enum=Append;Replace
+    Action IngressControllerHTTPHeaderAction `json:"action"`
+}
+```
+
+We will be using the exisiting API:
+
+```go
+// IngressControllerHTTPHeaderAction is a policy for setting HTTP headers.
+// +kubebuilder:validation:Enum=Append;SetOrReplace
+type IngressControllerHTTPHeaderAction string
+
+const (
+  // AppendHTTPHeader appends the header, preserving any existing header.
+  AppendHTTPHeader IngressControllerHTTPHeaderAction = "Append"
+  // SetOrReplaceHTTPHeader sets the header, removing any existing header.
+  SetOrReplaceHTTPHeader IngressControllerHTTPHeaderAction = "SetOrReplace"
+)
```

Comment:
Yeah, I think you have the point about the `Value` field. It's probably better to be validated by the OpenAPI, this may be clearer for the end user.
However, what about the actions? How do you plan to use them in the cluster-ingress-operator or router?

Reply:
Consider using a [discriminated union](https://github.com/openshift/enhancements/blob/master/dev-guide/api-conventions.md#discriminated-unions) where `IngressControllerHTTPHeaderAction` would be the discriminator, "Delete" would be an empty union member, and "Set" would be union member with a "Value" field.  "Replace" and "Append" could be expressed as separate union members or as a field of the "Set" union member.

Reply:
I can see the delete as a union member:
```go
// +union
type IngressControllerHTTPHeader struct {
  // +unionDiscriminator
  // +kubebuilder:validation:Required
  Action string `json:"action,omitempty"`
  // +optional.
  Set *IngressControllerSetHTTPHeader `json:"set,omitempty"`
  // +optional.
  Delete *IngressControllerDeleteHTTPHeader `json:"delete,omitempty"`
}
```
```yaml
apiVersion: operator.openshift.io/v1                                                                                                         
kind: IngressController
metadata:
  name: default
spec:
  httpHeaders:
    headerManipultation:
      response:
      - action: "Set"
        name: "X-Frame-Options"
        value: "DENY"
      - action: "Delete"
        name: "X-Powered-By"
```
but I cannot quite understand how it can be an empty union member.
Does it mean that only `Delete` action will be specified?

Reply:
For Action we can specify 

> Consider using a [discriminated union](https://github.com/openshift/enhancements/blob/master/dev-guide/api-conventions.md#discriminated-unions) where `IngressControllerHTTPHeaderAction` would be the discriminator, "Delete" would be an empty union member, and "Set" would be union member with a "Value" field. "Replace" and "Append" could be expressed as separate union members or as a field of the "Set" union member.

yeah it won't empty. IngressControllerDeleteHTTPHeader will need header name which needs to be passed.

Reply:
Thanks @Miciah for clarifying.
I'm almost fine with the discriminated union approach. It just feels a little verbose, especially on this part:
```yaml
- action: "Set"
  name: "X-Frame-Options"
  set:  <-- this part
    value: "DENY"
```

Reply:
> I'm almost fine with the discriminated union approach. It just feels a little verbose

True, but it's a standard pattern and makes it clear which fields apply to which action.

Reply:
Sorry to bring this back but I don't think I'm 100% convinced that the discriminated union approach is better than the dedicated Set and Delete fields: 
```yaml
apiVersion: operator.openshift.io/v1                                                                                                         
kind: IngressController
metadata:
  name: default
spec:
  httpHeaders:
    headerManipultation:
      response:
        set:
        - name: "X-Frame-Options"
          value: "DENY"
        delete:
        - name: "X-Powered-By"
```

The API maybe a little less redundant for the discriminated union (`Name` field is common, no `DeleteHeader` type is needed) but is it more convenient to the end user? IMO it feels a little more verbose than the variant I showed above.

Reply:
The discriminated union approach allows the API users to create an ordered list of the header manipulation instructions. This maybe more explicit expression of their intent as the dedicated set/delete fields approach would always raise the question in which order the two lists need to be translated into HAProxy configuration.

Reply:
> Btw don't we loose some validation using `Name` as an empty union member ([same question](https://github.com/openshift/enhancements/pull/1296#discussion_r1064629179))? it has to be optional while we need it required, no?

For the record, I believe I answered the question here: https://github.com/openshift/enhancements/pull/1296#discussion_r1064668220

> `Name` should not be a union member. You can use a inline field for the union type. See `IngressControllerCaptureHTTPCookie` ([source](https://github.com/openshift/api/blob/5ad311dde2264b659a256a17c27f1836fb50e85d/operator/v1/types_ingress.go#L1100-L1117)) for an example.

Please let me know if that doesn't satisfactorily answer your question!